ttnn/cpp/ttnn/operations/data_movement/tilize/device/tilize_op.cpp

```diff
@@ -0,0 +1,142 @@
+    const std::vector<Tensor>& input_tensors, const std::vector<std::optional<Tensor>>& output_tensors) const {
+    const auto& input_tensor = input_tensors.at(0);
+    if (input_tensor.memory_config().is_sharded()) {
+        auto mem_config = this->output_mem_config;
+        mem_config.shard_spec = input_tensor.memory_config().shard_spec;
+        return {create_device_tensor(
+            this->compute_output_shapes(input_tensors).at(0),
+            this->output_dtype,
+            Layout::TILE,
+            input_tensor.device(),
+            mem_config)};
+    } else {
+        return operation::generic_create_output_tensors(
+            *this, input_tensors, this->output_dtype, Layout::TILE, this->output_mem_config);
+    }
+}
+
+operation::ProgramWithCallbacks Tilize::create_program(
+    const std::vector<Tensor>& input_tensors, std::vector<Tensor>& output_tensors) const {
+    const auto& input_tensor_a = input_tensors.at(0);
+    auto& output_tensor = output_tensors.at(0);
+    if (this->use_multicore) {
+        return detail::tilize_multi_core(input_tensor_a, output_tensor);
+    }
+    return detail::tilize_single_core(input_tensor_a, output_tensor);
+}
+
+void TilizeWithValPadding::validate(const std::vector<Tensor>& input_tensors) const {
+    const auto& input_tensor_a = input_tensors.at(0);
+    TT_FATAL(input_tensor_a.storage_type() == StorageType::DEVICE, "Operands need to be on device!");
+    TT_FATAL(input_tensor_a.buffer() != nullptr, "Operands need to be allocated in buffers on device!");
+    TT_FATAL(input_tensor_a.get_layout() == Layout::ROW_MAJOR, "Can only tilize row major data");
+    TT_FATAL(input_tensor_a.get_dtype() == DataType::BFLOAT16);
+
+    TT_FATAL(input_tensor_a.get_legacy_shape()[0] <= this->output_tensor_shape[0]);
+    TT_FATAL(input_tensor_a.get_legacy_shape()[1] <= this->output_tensor_shape[1]);
+    TT_FATAL(input_tensor_a.get_legacy_shape()[2] <= this->output_tensor_shape[2]);
+    TT_FATAL(input_tensor_a.get_legacy_shape()[3] <= this->output_tensor_shape[3]);
+
+    uint32_t num_rows = this->output_tensor_shape[2];
+    uint32_t inner_dim = this->output_tensor_shape[3];
+    TT_FATAL(num_rows % TILE_HEIGHT == 0, "Output shape must be tilizable");
+    TT_FATAL(inner_dim % TILE_WIDTH == 0, "Output shape must be tilizable");
+
+    if (input_tensor_a.memory_config().is_sharded()) {
+        TT_FATAL(input_tensor_a.memory_config().memory_layout == TensorMemoryLayout::WIDTH_SHARDED);
+        TT_FATAL(this->output_mem_config.memory_layout == input_tensor_a.memory_config().memory_layout);
+        for (uint32_t i = 0; i < input_tensor_a.get_legacy_shape().rank(); i++) {
+            if (i != input_tensor_a.get_legacy_shape().rank() - 2) {
+                TT_FATAL(input_tensor_a.get_legacy_shape()[i] == this->output_tensor_shape[i]);
+            }
+        }
+    }
+}
+
+std::vector<tt::tt_metal::Shape> TilizeWithValPadding::compute_output_shapes(
```

Comment:
I think we should move this to a different file tilize_with_val_padding_op. 
Wdyt?